### ISSUE TITLE
Q. Get affinity matrix for random walk.

### ISSUE THREAD

**Issue by sunshower76** [ajabri/videowalk#33]
Hello. Thanks to your work!

I've referred to your code and have a question.

please see your code   **As = self.affinity(q[:, :, :-1], q[:, :, 1:])  (code/model.py, line 140.)**

we can define the affinity matrix that walk frame 1 to frame 2 as **torch.matmul(frame2, frame1)**
and then, the walk frame1 to frame3 could be gotten as   **matmul( matmul(frame3, frame2), matmul(frame2, frame1)  )**

As a result, I think your code should be changed **As = self.affinity(q[:, :, :-1], q[:, :, 1:]) ** to **As = self.affinity(q[:, :, 1:], q[:, :, :-1]) **

But, you got a good performance in your expriment. It seems that I miss some figures.
Could you explain it?

Thanks. :)

**Comment by ajabri** [comment]
Hi @sunshower76, if you read the code in `affinity`, we use `torch.einsum('bctn,bctm->btnm', q[:, :, :-1], q[:, :, 1:])` to compute affinities. This has the same effect.

**Comment by sunshower76** [comment]
Thank you for replying.

Is `torch.einsum('bctn,bctm->btnm', q[:, :, :-1], q[:, :, 1:])`  equal to 'matmul(matmul(frame n+1, frame n ))' ?
I think the code makes the result, `[matmaul(frame1, frame2), matmul(frame2, frame3) ...]`

I think that `torch.einsum('bctn,bctm->btnm', q[:, :, 1:], q[:, :, :-1])` is right.
As a result, affinities should make the result like this `matmaul(frame2, frame1), matmul(frame3, frame2) ...`

Thank you :)

**Comment by ajabri** [comment]
It depends on whether your transition matrices are row or column normalized (see https://en.wikipedia.org/wiki/Stochastic_matrix#Definition_and_properties). 

If they are row normalized, one should compute:
```
A_1^{t} = A_1^2 A_2^3... A_{t-1}^{t}
```

If they are column normalized, one should compute:
```
A_1^{t} = A_{t-1}^{t} ...  A_2^3 A_1^2 
```

In our code, we row normalize. Does this address your concern?

**Comment by sunshower76** [comment]
Thank you.
your reply address my concern exactly.

For example, when we warp the image source to style , in our case warping 'frame 1' to 'frame 2', we compute operation below.
` F_1^{2} = torch.matmul( F.softmax( torch.matmul(frame2, frame1), dim=-1), frame1 )`   
and then we can find that, the stochastic matrix is `A_1^{2} = F.softmax( torch.matmul(frame2, frame1) `  

Continuously, we can get approximate `A_2^{3} = F.sotmax( torch.matmul(frame3, frame2), dim=-1)` .
Following the above logic , `F_1^{3} =  torch.matmul( A_2^{3}, F_1^{2} )`  

As a result, I think this is the **row normalized case** but, the chaining order is the reverse to you mentioned.
Finally, I think the transition matrix should be computed in row normalized case:
`A_1^{t} = A_{t-1}^{t} ...  A_2^3 A_1^2`  

thank you :)

**Comment by ajabri** [comment]
Hi @sunshower76,

I think it depends on what shape `A` is. A simple way to see the correct ordering is if your affinities are not square, i.e. `A_1` is `d1 x d2` and `A_2` is `d2 x d3`. Then, you must have `torch.matmul(A_1, A_2)`.

Does this make sense?

Allan
